### PR TITLE
Recursion

### PR DESCRIPTION
**Pull request by memcmahon**

### CODE REVIEW

**Comment by memcmahon** [comment]
@zoefarrell I think this is now pretty final... I don't know that the lesson flows super well... but in a short amount of time I wanted to make sure to cover 'what is recursion' and how the stack plays into things (I did not really understand recursion until I understood that the stack was a thing 😆 ).  I don't know that we will have time to get to the exercises, but I wanted to include them for folks that wanted to try their hand at some practice.

File: module6/lessons/Week5/Recursion.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		while (num > -1)
+		{
+			Console.WriteLine(num);
+			num--;
+		}
+	}
+	
+	static void RecursiveCountdown(int num)
+	{
+		if (num == )
```
Comment (zoefarrell) [inline module6/lessons/Week5/Recursion.md]: I like this example showing looping vs recursion! Does this work with nothing after the ==?

File: module6/lessons/Week5/Recursion.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		else
+		{
+			Console.WriteLine(num);
+			RecursiveCountdown(num - 1);
+		}
+	}
+}
+```
+
+This is a very simple example of recursion, but it does give us a chance to identify the key features of a recursive function.  Every recursive function must have these two pieces:
```
Comment (zoefarrell) [inline module6/lessons/Week5/Recursion.md:68]: I'm glad that you have these two parts written out! I find always coming back to these two things is helpful when I've taught recursion in the past.

File: module6/lessons/Week5/Recursion.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+Reverse a string.
+
+```js
+// create a function which takes a string of characters and
+// recursively calls itself to reverse the string
+
+// e.g.
+
+string reversedString = reverse('Ariel');
```
Comment (zoefarrell) [inline module6/lessons/Week5/Recursion.md:135]: I might add an instructor note here to talk as a class through the base case and the recursive case (how does the recursive case get closer to the base case). Or maybe do a breakout room to talk with your partner and then write the pseudo code as a class. I think starting from pseudocode for the first one will be helpful!

File: module6/lessons/Week5/Recursion.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+Use the slides below to walkthrough what is happening on the stack.  Not _everything_ that is happening on the stack is specifically called out - I tried to include only the most necessary information.
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### Diagramming the Call Stack
+
+<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vRFJ2pNfsOap3jgooinxMqFWbcKASjPv8b0TY_lHCh53vD3ts0DE_fGqQuJclIaWkctq-qfLtZjJ0r0/embed?start=false&loop=false&delayms=60000" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>
```
Comment (zoefarrell) [inline module6/lessons/Week5/Recursion.md:101]: These slides are awesome!!

**Comment by zoefarrell** [comment]
> @zoefarrell I think this is now pretty final... I don't know that the lesson flows super well... but in a short amount of time I wanted to make sure to cover 'what is recursion' and how the stack plays into things (I did not really understand recursion until I understood that the stack was a thing 😆 ). I don't know that we will have time to get to the exercises, but I wanted to include them for folks that wanted to try their hand at some practice.

I agree! This lesson seems great! I think getting to coding the first reverse a string example would be great and then have the others as resources we reference back to for freethrow time.

**Review by zoefarrell — APPROVED** [review]

File: module6/lessons/Week5/Recursion.md (hunk 1 above)
Comment (memcmahon) [reply module6/lessons/Week5/Recursion.md]: haha - definitely not 😆

**Pull request merged**